### PR TITLE
EPROD-992 pagination for icrc2 operation list queries

### PR DESCRIPTION
**Pull request by veeso**
## Issue ticket

Issue ticket link:

### Code conventions

- [ ] I have performed a self-review of my code
- [ ] Every new function is documented
- [ ] Object names are auto explicative
- [ ] The PR follows the [project conventions](https://infinityswap.atlassian.net/wiki/spaces/CPROD/pages/23330839/Conventions)
- [ ] The code follows the [style guidelines of the project](https://infinityswap.atlassian.net/wiki/spaces/CPROD/pages/24444929/Rust+Conventions)
- [ ] The code follows the [project security best practices](https://infinityswap.atlassian.net/wiki/spaces/CPROD/pages/35225620/Security+Considerations)

### Security 

- [ ] The PR does not break APIs backward compatibility
- [ ] The PR does not break the stable storage backward compatibility

### Testing 

- [ ] Every function is properly unit tested
- [ ] I have added integration tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] IC endpoints are always tested through the `canister_call!` macro

### CODE REVIEW

File: src/icrc2-minter/src/canister.rs [hunk 1]
```diff
@@ -388,6 +392,28 @@ impl MinterCanister {
     pub fn idl() -> Idl {
         generate_idl!()
     }
+
+    /// Get mint orders for the given wallet address and token;
+    /// if `offset` and `count` are provided, returns a page of mint orders.
+    fn token_mint_orders(
+        wallet_address: H160,
+        src_token: Id256,
+        offset_and_count: Option<(usize, usize)>,
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:401]: usually, paginated APIs allow specifying offset and count independently

File: src/icrc2-minter/src/canister.rs [hunk 2]
```diff
@@ -388,6 +392,28 @@ impl MinterCanister {
...
+    ) -> Vec<(u32, SignedMintOrder)> {
+        match offset_and_count {
+            Some((offset, count)) => {
+                get_operations_store().get_page_for_address(&wallet_address, offset, count)
+            }
+            None => get_operations_store().get_for_address(&wallet_address),
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:407]: Do not duplicate the `get_for_address` fn

File: src/icrc2-minter/src/constant.rs [hunk 3]
```diff
@@ -12,3 +12,5 @@ pub const OPERATIONS_MAP_MEMORY_ID: MemoryId = MemoryId::new(90);
 pub const DEFAULT_TX_GAS_LIMIT: u64 = 3_000_000;
 
 pub const IC_CHAIN_ID: u32 = 0;
+
+pub const MAX_OPERATIONS_LIST_SIZE: usize = 1024;
```
Comment (ufoscout) [inline src/icrc2-minter/src/constant.rs:16]: This is used for minted orders, the name is confusing.
Add a description

File: src/integration-tests/tests/pocket_ic_integration_test/icrc2_minter.rs [hunk 4]
```diff
@@ -191,7 +191,10 @@ async fn test_icrc2_token_canister_stopped() {
 
     let (_, refund_mint_order) = ctx
         .client(ctx.canisters().icrc2_minter(), ADMIN)
-        .query::<_, Vec<(u32, SignedMintOrder)>>("list_mint_orders", (john_address, base_token_id))
+        .query::<_, Vec<(u32, SignedMintOrder)>>(
```
Comment (ufoscout) [inline src/integration-tests/tests/pocket_ic_integration_test/icrc2_minter.rs:194]: Add the fn to the client and use it here

File: src/minter-contract-utils/src/operation_store.rs [hunk 5]
```diff
@@ -270,6 +273,27 @@ where
             log::trace!("Operation {id} is evicted from the operation log");
         }
     }
+
+    /// Retrieves all operations for the given ETH wallet address, starting from `offset` returning a max of `count` items
+    /// If `offset` is `None`, it starts from the beginning.
+    /// If `count` is `None`, it returns all operations.
+    fn operations_for_address(
```
Comment (ufoscout) [inline src/minter-contract-utils/src/operation_store.rs:280]: This fn covers all use cases. Remove `get_page_for_address` and `get_for_address`

File: src/minter-contract-utils/src/operation_store.rs [hunk 6]
```diff
@@ -352,6 +376,28 @@ mod tests {
         }
     }
 
+    #[test]
+    fn should_get_page_for_operations() {
```
Comment (ufoscout) [inline src/minter-contract-utils/src/operation_store.rs:380]: Add additional tests for all combinations of offset and count (e.g. Some/Some, Some/None, None/Some, None/None)

File: src/icrc2-minter/src/canister.rs [hunk 7]
```diff
@@ -276,16 +276,14 @@ impl MinterCanister {
         &self,
         wallet_address: H160,
         src_token: Id256,
+        offset: usize,
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:279]: should these args be Optional?

File: src/icrc2-minter/src/canister.rs [hunk 8]
```diff
@@ -306,8 +304,14 @@ impl MinterCanister {
     pub fn get_operations_list(
         &self,
         wallet_address: H160,
+        offset: usize,
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:307]: should these args be Optional?

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/icrc2-minter/src/canister.rs [hunk 9]
```diff
@@ -388,6 +389,25 @@ impl MinterCanister {
...
+    fn token_mint_orders(
+        wallet_address: H160,
+        src_token: Id256,
+        offset: Option<usize>,
+        count: Option<usize>,
+    ) -> Vec<(u32, SignedMintOrder)> {
+        get_operations_store()
+            .get_for_address(&wallet_address, offset, count)
+            .into_iter()
+            .filter_map(|(operation_id, status)| {
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:404]: This filter is applied after the offset and count, consequently the offset of the returned result and the number of returned entries do not match what requested

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/icrc2-minter/src/canister.rs [hunk 10]
```diff
@@ -388,6 +389,27 @@ impl MinterCanister {
...
+    ) -> Vec<(u32, SignedMintOrder)> {
+        get_operations_store()
+            .get_for_address(&wallet_address, None, None)
+            .into_iter()
+            .filter_map(|(operation_id, status)| {
+                status
+                    .get_signed_mint_order(Some(src_token))
+                    .map(|mint_order| (operation_id.nonce(), *mint_order))
+            })
+            .skip(offset.unwrap_or_default())
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:409]: After this change, the two new Optional parameters to the `get_for_address` fn are never used.

File: src/icrc2-minter/src/canister.rs [hunk 11]
```diff
@@ -388,6 +389,27 @@ impl MinterCanister {
     pub fn idl() -> Idl {
         generate_idl!()
     }
+
+    /// Get mint orders for the given wallet address and token;
+    /// if `offset` and `count` are provided, returns a page of mint orders.
+    fn token_mint_orders(
```
Comment (ufoscout) [inline src/icrc2-minter/src/canister.rs:395]: This fn has not tested filtering logic , please add proper testing

**Review by ufoscout — CHANGES_REQUESTED** [review]

File: src/icrc2-minter/src/canister.rs (hunk 10 above)
Comment (veeso) [reply src/icrc2-minter/src/canister.rs:409]: #[query]
    /// Returns the list of operations for the given wallet address.
    /// Offset, if set, defines the starting index of the page,
    /// Count, if set, defines the number of elements in the page.
    pub fn get_operations_list(
        &self,
        wallet_address: H160,
        offset: Option<usize>,
        count: Option<usize>,
    ) -> Vec<(MinterOperationId, OperationState)> {
        get_operations_store().get_for_address(&wallet_address, offset, count)
    }

**Review by ufoscout — APPROVED** [review]

**Pull request merged**